tests/core/test_decorators.py

```diff
@@ -0,0 +1,33 @@
+import pytest
+import torch
+
+from tests.base import EvalModelTemplate
+from pytorch_lightning.core.decorators import auto_move_data
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
+@pytest.mark.parametrize(['src_device', 'dest_device'], [
+    pytest.param(torch.device('cpu'), torch.device('cpu')),
+    pytest.param(torch.device('cpu', 0), torch.device('cuda', 0)),
+    pytest.param(torch.device('cuda', 0), torch.device('cpu')),
+    pytest.param(torch.device('cuda', 0), torch.device('cuda', 0)),
+])
+def test_auto_move_data(src_device, dest_device):
+    """ Test that the decorator moves the data to the device the model is on. """
+
+    class CurrentModel(EvalModelTemplate):
+        pass
+
+    # apply the decorator
+    CurrentModel.forward = auto_move_data(CurrentModel.forward)
+
+    model = CurrentModel()
```

Comment:
```suggestion
    model = EvalModelTemplate()
    # apply the decorator
    model.forward = auto_move_data(model.forward)
```
why pure `EvalModelTemplate` is not enough and you need a child?

Reply:
unfortunately not working. the decorator needs to be applied to the class definition. 
So we can do `EvalModelTemplate.forward = auto_move_data(EvalModelTemplate.forward)` but then it is applied for all following tests too.

Reply:
yeah, that looks easier :rabbit:

Reply:
this has side effects on other tests (and are failing). the decorator should only be applied at test time, when we e.g. only have one gpu. 
I think it's cleaner if we applied the decorator only locally in this test.